### PR TITLE
Fujitv premium info extract

### PR DESCRIPTION
**Pull request by YuenSzeHong**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

### CODE REVIEW

**Comment by pukkandan** [comment]
Depends on #1969, correct?

File: yt_dlp/extractor/fujitv.py [hunk 1]
```diff
@@ -5,44 +5,67 @@
...
+    _VALID_URL = r'https?://fod\.fujitv\.co\.jp/title/(?P<sid>[0-9a-z]{4})/(?P<id>[0-9a-z]+)'
+    _BASE_URL = 'https://i.fod.fujitv.co.jp/'
+    # _BITRATE_MAP = {
+    #     300: (320, 180),
+    #     800: (640, 360),
+    #     1200: (1280, 720),
+    #     2000: (1280, 720),
+    #     4000: (1920, 1080),
+    #     6000: (1920, 1080),
+    # }
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:17]: Don't leave commented code. If it's not used, delete it. If it's stuff to fix, mark PR as draft

**Comment by YuenSzeHong** [comment]
> Depends on #1969, correct?

yes, info extraction code from #1969

File: yt_dlp/extractor/fujitv.py [hunk 2]
```diff
@@ -5,44 +5,48 @@
...
-        self._sort_formats(formats)
+        series_id, video_id = self._match_valid_url(url).groups()
+        self._download_webpage(url, video_id)
+        json_info = {}
+        token = self._get_cookies(url).get('CT')
+        if token:
+            json_info = self._download_json('https://fod-sp.fujitv.co.jp/apps/api/episode/detail/?ep_id=%s&is_premium=false' % video_id, video_id, headers={'x-authorization': f'Bearer {token.value}'}, fatal=False)
+        else:
+            self.report_warning('Unable to extract token cookie, video information is unavailable')
+            print(self._LOGIN_HINTS['cookies'])
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:33]:
```suggestion
            self.report_warning(f'The token cookie is needed to extract video metadata. {self._LOGIN_HINTS["cookies"]}')
```

File: yt_dlp/extractor/fujitv.py [hunk 3]
```diff
@@ -5,44 +5,48 @@
...
             'id': video_id,
-            'title': video_id,
+            'title': json_info.get('ep_title'),
+            'series': json_info.get('lu_title'),
+            'series_id': series_id,
+            'description': json_info.get('ep_description'),
             'formats': formats,
-            'thumbnail': self._BASE_URL + 'pc/image/wbtn/wbtn_%s.jpg' % video_id,
+            'subtitles': subtitles,
+            'thumbnail': self._BASE_URL + f'img/program/{series_id}/episode/{video_id}_a.jpg',
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:51]:
```suggestion
            'thumbnail': f'{self._BASE_URL}img/program/{series_id}/episode/{video_id}_a.jpg',
```

File: yt_dlp/extractor/fujitv.py [hunk 4]
```diff
@@ -5,44 +5,48 @@
...
+        else:
+            self.report_warning('Unable to extract token cookie, video information is unavailable')
+            print(self._LOGIN_HINTS['cookies'])
+        formats = []
+        subtitles = []
+        src_json = self._download_json(self._BASE_URL + 'abrjson_v2/tv_android/%s' % video_id, video_id)
+        for src in src_json.get('video_selector'):
+            format, subtitle = self._extract_m3u8_formats_and_subtitles(src.get('url'), video_id, 'mp4')
+            formats += format
+            subtitles += subtitle
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:40]:
```suggestion
        formats, subtitles = [], {}
        src_json = self._download_json(f'{self._BASE_URL}abrjson_v2/tv_android/{video_id}', video_id)
        for src in src_json['video_selector']:
            if not src.get('url'):
                continue
            fmt, subs = self._extract_m3u8_formats_and_subtitles(src['url'], video_id, 'mp4')
            formats.extend(fmt)
            subtitles = self._merge_subtitles(subtitles, subs)
```

File: yt_dlp/extractor/fujitv.py [hunk 5]
```diff
@@ -5,44 +5,48 @@
...
+        self._sort_formats(formats, ['tbr'])
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:41]: why tbr?

File: yt_dlp/extractor/fujitv.py [hunk 6]
```diff
@@ -5,44 +5,48 @@
...
-        for f in formats:
-            wh = self._BITRATE_MAP.get(f.get('tbr'))
-            if wh:
-                f.update({
-                    'width': wh[0],
-                    'height': wh[1],
-                })
-        self._sort_formats(formats)
+        series_id, video_id = self._match_valid_url(url).groups()
+        self._download_webpage(url, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:26]: check if a `HEADRequest` is sufficient

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/fujitv.py [hunk 7]
```diff
@@ -5,44 +5,48 @@
...
-        video_id = self._match_id(url)
-        formats = self._extract_m3u8_formats(
-            self._BASE_URL + 'abr/tv_android/%s.m3u8' % video_id, video_id, 'mp4')
-        for f in formats:
-            wh = self._BITRATE_MAP.get(f.get('tbr'))
-            if wh:
-                f.update({
-                    'width': wh[0],
-                    'height': wh[1],
-                })
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:40]: Why is this removed?

File: yt_dlp/extractor/fujitv.py (hunk 5 above)
Comment (YuenSzeHong) [reply yt_dlp/extractor/fujitv.py:41]: because some formats has no resolution, so screwing up the sorting

File: yt_dlp/extractor/fujitv.py (hunk 7 above)
Comment (YuenSzeHong) [reply yt_dlp/extractor/fujitv.py:40]: no need for this, the JSON returns the main profile, which has resolution info in it

Comment (pukkandan) [reply yt_dlp/extractor/fujitv.py:40]: You are contradicting what you said here: https://github.com/yt-dlp/yt-dlp/pull/2505#discussion_r794426783

Comment (YuenSzeHong) [reply yt_dlp/extractor/fujitv.py:40]:
```json
{
    "video_selector": [
        {
            "value": 0,
            "title": "自動",
            "url": "https://hls-vod-auth.stream.co.jp/hls-vod-auth/fod-wse-abr/meta.m3u8?tk=2bbb6cafae72562513f8e14641da32d63326eb8b85824a533f0415e45aba86cc"
        },
        {
            "value": 1,
            "title": "低",
            "url": "https://hls-vod-auth.stream.co.jp/hls-vod-auth/fod-wse-mid/meta.m3u8?tk=22ca4082f1c43a1779212cc9ae7240ba0a49fd9c3213e5179e5150b801c77300cc9e501531d180ebc7fc6e76fd837439"
        },
        {
            "value": 2,
            "title": "高",
            "url": "https://hls-vod-auth.stream.co.jp/hls-vod-auth/fod-wse-high/meta.m3u8?tk=22ca4082f1c43a1779212cc9ae7240ba716c7e9ae4c18c7c6845e1b607c91eb297c10a1333730b525fbf255282b4dbfc"
        }
    ]
}
```
the 低 and 高 one return formats without resolution and the 自動 return formats with a resolution, and it recons the peak bitrate, so impossible to map

File: yt_dlp/extractor/fujitv.py (hunk 6 above)
Comment (YuenSzeHong) [reply yt_dlp/extractor/fujitv.py:26]: image
seems ok

File: yt_dlp/extractor/fujitv.py [hunk 8]
```diff
@@ -1,48 +1,52 @@
...
-        self._sort_formats(formats)
+        series_id, video_id = self._match_valid_url(url).groups()
+        self._download_webpage(Request(url, method='head'), video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/fujitv.py:26]:
```suggestion
        self._request_webpage(HEADRequest(url), video_id)
```
import `HEADRequest` from utils

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/fujitv.py (hunk 8 above)
Comment (YuenSzeHong) [reply yt_dlp/extractor/fujitv.py]: which utils? this?
image
